pkg/server/user_connections.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2025 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"context"
+
+	"github.com/pingcap/tidb/pkg/privilege"
+	servererr "github.com/pingcap/tidb/pkg/server/err"
+	"github.com/pingcap/tidb/pkg/sessionctx/vardef"
+	"github.com/pingcap/tidb/pkg/util/logutil"
+	"go.uber.org/zap"
+)
+
+// incrementUserConnectionsCounter increases the count of connections when user login the database.
+func (cc *clientConn) incrementUserConnectionsCounter() {
+	user := cc.ctx.GetSessionVars().User
+	targetUser := user.AuthUsername + user.AuthHostname
```

Comment:
we should escape and concatenate them. I'm not sure if there's a case that user = `a` host = `bc` and user = `ab` host = `c`

Reply:
Modified it.